torch/jit/_check.py

```diff
@@ -0,0 +1,214 @@
+        # tuple in `__init__` CORRECTLY means that that variable
+        # cannot be reassigned later to a non-empty tuple. Same
+        # deal with `NamedTuple`
+
+        containers = {"List", "Dict", "Optional"}
+
+        # If we're not evaluating one of the specified problem types
+        try:
+            if node.annotation.value.id not in containers:
+                return
+        except AttributeError:
+            # To evaluate a base type (`str`, `int`, etc.), we would
+            # have needed to get the name through `node.annotation.id`
+            # instead of `node.annotation.value.id`. Seems that we're
+            # not evaluating one of our "containers"
+            return
+
+        # switch-case to check if the assigned variable is not empty
+        ann_type = node.annotation.value.id
+        if ann_type == "List":
+            # Assigning `[]` to a `List` type gives you an AnnAssign
+            # Node where value=List(elts=[], ctx=Load())
+            if not isinstance(node.value, ast.List):
+                return
+            if node.value.elts:
+                return
+        elif ann_type == "Dict":
+            # Assigning `{}` to a `Dict` type gives you an AnnAssign
+            # Node where value=Dict(keys=[], values=[])
+            if not isinstance(node.value, ast.Dict):
+                return
+            if node.value.keys:
+                return
+        elif ann_type == "Optional":
+            # Assigning `None` to an `Optional` type gives you an
+            # AnnAssign Node where value=Constant(value=None, kind=None)
+            # or, in Python <3.8, value=NameConstant(value=None)
+            if (not self.using_deprecated_ast
+                    and not isinstance(node.value, ast.Constant)):
+                return
+            if (self.using_deprecated_ast
+                    and not isinstance(node.value, ast.NameConstant)):
+                return
+            if node.value.value:
+                return
+
+        raise RuntimeError("The TorchScript type system doesn't support"
+                           " instance-level annotations on empty"
+                           " non-base types in `__init__`. Instead, "
+                           " either 1) use a type annotation in the "
+                           "class body, or 2) wrap the type in "
+                           "`torch.jit.Attribute`.")
+
+    def visit_Call(self, node):
+        """
+        Visit a Call node in an ``nn.Module``'s ``__init__``
+        method and determine if it's ``torch.jit.annotate``. If so,
+        see if it conforms to our attribute annotation rules.
+        """
+        # If this isn't a call to `torch.jit.annotate`
+        try:
+            if (node.func.value.value.id != "torch"
+                    or node.func.value.attr != "jit"
+                    or node.func.attr != "annotate"):
+                try:
+                    self.generic_visit(node)
+                except RuntimeError:
```

Comment:
Shouldn't we just call `self.generic_visit` and let the error propagate naturally? What is the purpose of the try-except-raise?